---
layout: page
title: "RNASeq Training"
permalink: /rnaseq/
tags: [ rnaseq, transcriptomics ]
description: RNASeq page
---
<table class="table-contact">
<tr>
<td><img width="100%" src="{{ site.url }}/images/trainings-rnaseq.png" alt="" />
</td>
<td>
<h1> Introduction to RNASeq analysis </h1><br />
This course offers an introduction to RNASeq analysis. We will cover from raw reads to differential expression analysis.
</td>
</tr>
</table>

### Prerequisites
Linux for dummies
<div id="colonne1">
<h3>Program</h3>
<ul>
<li> Reads preprocessing </li>
<li> Mapping RNASeq reads against an annotated reference genome with STAR </li>
<li> Counting reads in transcrits/genes with HTSeq-count </li>
<li> Count normalization using Diane </li>
<li> Differential expression analysis using Diane </li>
<li> Plots, clustering, co-expression network: degust, WGCNA </li>
</ul>
</div>

<div id="colonne2">
<h3>Learning objectives</h3>
<ul>
<li>Process raw reads to count matrix </li>
<li>Think about different normalisation methods</li>
<li>Detect differentially expressed genes</li>
<li>Compare results between approaches</li>
</ul>
</div>

<div id="colonne3">
<h3>Training material</h3>
<ul>
<li>Slides : <a target="_blank" href="{{ site.url }}files/analyse_rnaseq_2022.pdf">see</a></li>
<li>Practice : <a target="_blank" href="{{ site.url }}/linux/rnaseqPractice">see</a> </li>
</ul>
</div>

<div id="nextInline" class="clearfix">
<h3>Instructors</h3>
<ul>
    <li>Guilhem Sempéré (GS) - guilhem.sempere@cirad.fr</li>
    <li>Alexandre Sorianon (AS) - alexandre.soriano@cirad.fr</li>
    <li>Catherine Breton (CB) - c.breton@cgiar.org</li>
    <li>Gautier Sarah (GS) - gautier.sarah@cirad.fr</li>
    <li>Julie Orjuela (JO) - julie.orjuela@ird.fr</li>
    <li>Aurore Comte (AC) - aurore.compte@ird.fr</li>
    <li>Christine Tranchant (CT) - christine.tranchant@ird.fr</li>
    <li>Sebastien Ravel (SR) - sebastien.ravel@cirad.fr </li>
    <li>Alexis Dereeper (AD) - alexis.dereeper@ird.fr </li>
    <li>Sebastien Cunnac (SC) - sebastien.cunnac@ird.fr </li>
</ul>
</div>

### Trainings
 
| **Date** | **Location** | **Topics** | **Parrticipants** | **Instructors** | **Links** | **Units** |
| 2023 | Montpellier, France |  Introduction to transcriptomics analysis | 15 | CB, GS, AS, GS | | UMR DIADE, AGAP, MIVEGEC, PHIM |
| 2022 | Montpellier, France |  Introduction to transcriptomics analysis | 10 | CB, GS, AS | | UMR DIADE, AGAP, MIVEGEC, PHIM |
| 2019 | Montpellier, France |  Introduction to transcriptomics analysis | 20 | JO, CB, GS, SC, AC | | UMR DIADE, IPME, AGAP, Bioversity  |
| 2018 | Montpellier, France |  Introduction to transcriptomics analysis | 20 | CT, SR, AD, SC | | UMR DIADE, IPME, BGPI |
| 2016 | Ouagadougou, Burkina |  Bioinformatics applied to genomics | 20 | CT, AD | | IRD |
| 2015 | Hanoi, Vietnam |  Bioinformatics applied to genomics | 20 | AD | | LMI Rice |
| 2014 | Dakar, Senegal |  Bioinformatics applied to genomics | 20 | CT, AD | | IRD, Bel Air Campus |

