---
layout: page
title: "RNASeq Training"
permalink: /rnaseq/
tags: [ rnaseq, transcriptomics ]
description: RNASeq page
---
<table class="table-contact">
<tr>
<td><img width="100%" src="{{ site.url }}/images/trainings-linux.png" alt="" />
</td>
<td>
<h1> RNASeq analyses (using Galaxy and TOGGLe)</h1><br />
This course offers an introduction to RNASeq analyses using two different workflow management systems: Galaxy and TOGGLe. This includes reference-based mapping, estimates of transcript levels, differential expression (DE) analyses, visualization of statistics results.
</td>
</tr>
</table>

### Prerequisites
Workflow management system (Galaxy, TOGGLe)
<div id="colonne1">
<h3>Program</h3>
<ul>
<li> Mapping of RNASeq against a transcriptome reference with kallisto (Galaxy) </li>
<li> Mapping of RNASeq against an annotated genome reference with TopHat (TOGGLe) </li>
<li> Differential expression analysis using EdgeR and DESeq2 </li>
<li> Plots, clustering, co-expression network: degust, WGCNA</li>
</ul>
</div>

<div id="colonne2">
<h3>Learning objectives</h3>
<ul>
<li>Manipulate packages/tools available for searching DE genes </li>
<li>Think about different normalisation methods</li>
<li>Detect differentially expressed genes</li>
<li>Compare results between two approaches</li>
</ul>
</div>

<div id="colonne3">
<h3>Training material</h3>
<ul>
<li>Slides : <a target="_blank" href="{{ site.url }}/files/analyse_rnaseq.pdf">see</a></li>
<li>Practice : <a target="_blank" href="{{ site.url }}/linux/rnaseqPractice">see</a> </li>
</ul>
</div>

<div id="nextInline" class="clearfix">
<h3>Instructors</h3>
<ul>
    <li>Christine Tranchant (CT) - christine.tranchant@ird.fr</li>
    <li>Sebastien Ravel (SR) - sebastien.ravel@cirad.fr </li>
    <li>Alexis Dereeper (AD) - alexis.dereeper@ird.fr </li>
    <li>Sebastien Cunnac (SC) - sebastien.cunnac@ird.fr </li>
</ul>
</div>

### Trainings
 
| **Date** | **Location** | **Topics** | **Parrticipants** | **Instructors** | **Links** | **Units** |
| 2018 | Montpellier, France |  Introduction to transcriptomics analyses | 20 | CT, SR, AD, SC | | UMR DIADE, IPME, BGPI |
| 2016 | Ouagadougou, Burkina |  Bioinformatics applied to genomics | 20 | CT, AD | | IRD |
| 2015 | Hanoi, Vietnam |  Bioinformatics applied to genomics | 20 | AD | | LMI Rice |
| 2014 | Dakar, Senegal |  Bioinformatics applied to genomics | 20 | CT, AD | | IRD, Bel Air Campus |

