---
layout: page
title: "SV Training"
permalink: /sv/
tags: [short and long reads, mapping, snp, SV]

description: SV page
---
<table class="table-contact">
<tr>
<td width="15%"><img width="80%" src="{{ site.url }}/images/trainings-ont.png" alt="" />
</td>
<td>
<h1> Introduction to Structural variant detection analyses</h1><br />
This course offers an introduction SNP and SV detection data analysis. 

The training program includes:

* Mapping of reads (short and long) against a reference genome (bwa, minimap2)

* SNP detection from mapping of short reads against a reference genome: SNP calling, filters and SNP annotation. Examples of possible studies based on SNP arrays

* Detection of SV by mapping of sequencing data ("short read" and "long read") against a reference genome (breakdancer, sniffle)

* SV detection from genome assembly and comparison (minimap2, nucmer, assemblytics, siry)

</td>
</tr>
</table>

### Prerequisites
Linux and knowledge of NGS formats 
<div id="colonne1">
<h3>Program</h3>
<ul>
<li>Introduction to the principles of the sequencing technology and its applications</li>
<li>Handling mapping tools suitable for ILLUMINA and ONT data</li>
<li>SNP calling with GATK</li>
<li>Detecting Structural Variations (SV) in short and long reads</li>
<li>SV detection from genome assembly and comparison (minimap2, nucmer, assemblytics, siry) </li>
</ul>
</div>


<div id="colonne2">
<h3>Learning objectives</h3>
<ul>
<li>Understanding limits and advantages of sequencing technology for SV detection</li>
<li>Manipulating short and long reads data on a virtual machine on jupyter environment</li>
<li>Handling mapping, SNP calling and SV detection tools and be able to analyse your own data</li> 
<li>Detecting structural variations comparing assemblies. </li>

</ul>
</div>


<div id="colonne3">
<h3>Training material</h3>
<ul>
<li>Slides : <a target="_blank" href="https://github.com/SouthGreenPlatform/training_SV_teaching/tree/2022/Topo/">see</a></li>
<li>Practice : <a target="_blank" href="https://github.com/SouthGreenPlatform/training_SV_teaching/tree/2022">see</a> </li>
</ul>
</div>

<div id="nextInline" class="clearfix">
<h3>Instructors</h3>
<ul>
    <li>Christine Tranchant (CT) - christine.tranchant@ird.fr </li>
    <li>Francois Sabot (FS) - francois.sabot@ird.fr </li>
    <li>Julie Orjuela (JO) - julie.orjuela@ird.fr</li>
    <li>Alexis Dereeper (AD) - alexis.dereeper@ird.fr </li>
</ul>
</div>

### Trainings
 
| **Date** | **Location** | **Topics** | **Participants** | **Instructors** | **Links** | **Units** |
| 2022 | Montpellier, France |  Introduction to SV detecton analyses | 15 | CT, FS, JO, AD | | UMR DIADE, PHIM  |
