---
layout: page
title: "Assembling and functionally annotating a de-novo transcriptome from RNA-seq data"
permalink: /trinity/
tags: [assembly, de novo, annotation ]
description: Trinity-Trinotate page
---
<table class="table-contact">
<tr>
<td width="15%"><img width="80%" src="{{ site.url }}/images/TrinotateLogo.png" alt="" />
</td>
<td>
<h1> Assembling and functionally annotating a de-novo transcriptome from RNA-seq data </h1><br />
This course offers an introduction to pre-processing, assembling and functionally annotating a de-novo transcriptome from RNA-seq data of non-model organisms.
</td>
</tr>
</table>

### Prerequisites
[Linux](https://southgreenplatform.github.io/trainings/linux/)
[LinuxJedi] (https://southgreenplatform.github.io/trainings//linuxJedi/)
[HPC] (/southgreenplatform.github.io/trainings//HPC/)
[RNAseq](https://southgreenplatform.github.io/trainings/rnaseq/)

<div id="colonne1">
<h3>Program</h3>
<ul>
<li> Transcrits assembly with Trinity  </li>
<li> Checking assembly quality  </li>
<li> Differential expression analysis of transcrits (DE)  </li>
<li> Funtional annotation of transcrits with Trinotate </li> 
<li> Codants regions prediction with TransDecoder </li>
</ul>
</div>

<div id="colonne2">
<h3>Learning objectives</h3>
<ul>
<li> Manipulate tools available for assemblies analysis </li>
<li> Study differential expression by using RNAseq data without a reference genome </li>
<li> Visualize diversity metrics in assemblies approaches </li>
<li> Annotate transcrits to know functions of genes  </li>
</ul>
</div>


<div id="colonne3">
<h3>Training material</h3>
<ul>
<li>Basic commands in SLURM : <a target="_blank" href="{{ site.url }}/slurm/">see</a></li>   
<li>Reads quality slides : <a target="_blank" href="{{ site.url }}/files/cleaning_2019.pdf">see</a></li>
<li>Practice 1. Checking Reads Quality <a target="_blank" href="{{ site.url }}/trinityTrinotate/TP-trinity/#practice-1">see</a> </li>
<li>Assembly slides : <a target="_blank" href="{{ site.url }}/files/trinity_2019.pdf">see</a></li>
<li>Practice 2. Assembling transcriptome from RNA-seq <a target="_blank" href="{{ site.url }}/trinityTrinotate/TP-trinity/#practice-2">see</a> </li>
<li>Practice 3. Assessing transcriptome assembly quality <a target="_blank" href="{{ site.url }}/trinityTrinotate/TP-trinity/#practice-3">see</a> </li>
<li>Annotation slides : <a target="_blank" href="{{ site.url }}/files/trinotate_2019.pdf">see</a></li>
<li>Practice 5 : Functional annotation <a target="_blank" href="{{ site.url }}/trinityTrinotate/TP-annotation">see</a> </li>
<li>Differential Expression slides : <a target="_blank" href="{{ site.url }}/files/de_2019.pdf">see</a></li>
<li>Practice 4 : Differential Expression Analysis (DE) <a target="_blank" href="{{ site.url }}/trinityTrinotate/TP-trinity/#practice-4">see</a> </li>
</ul>
</div>

<div id="nextInline" class="clearfix">
<h3>Instructors</h3>
<ul>
    <li> Erwan Corre (EC) - corre@sb-roscoff.fr </li>
    <li> Julie Orjuela-Bouniol (JO) - julie.orjuela@ird.fr</li> 
    <li> Christine Tranchant-Dubreuil (AD) - christine.tranchant@ird.fr </li>
    <li> Aurore Comte (AC) - aurore.comte@ird.fr</li>
    <li> Stephanie SIDIBE BOCS (SSB) -  stephanie.sidibe-bocs@cirad.fr </li>
    <li> David Lopez (DL) -  david.lopez@cirad.fr </li>
</ul>
</div>

### Trainings
 
| **Date** | **Location** | **Topics** | **Participants** | **Instructors** | **Links** | **Units** |
| 2019 | Montpellier, France |  Introduction to assembly and annotation of transcrits | 15 | EC JO CT AC SSB DL | | SOUTHGREEN ABiMS - IFB aliance|
