<!doctype html>
<!--[if lt IE 7]><html class="no-js ie6 oldie" lang="ja"><![endif]-->
<!--[if IE 7]>   <html class="no-js ie7 oldie" lang="ja"><![endif]-->
<!--[if IE 8]>   <html class="no-js ie8 oldie" lang="ja"><![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="ja"><!--<![endif]-->
<head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge,chrome=1">

    <title>Metabarcoding Practice</title>
    <meta name="description" content="Bioinformatics trainings, Linux, HPC, NGS, NGS workflow manager, TOGGLe, galaxy">
    <meta name="description" content="Metabarcoding Practice page">
    
    <meta name="author" content=" Sébastien RAVEL -- sebastien.ravel@cirad.fr">
    
    <meta name="author" content=" Christine Tranchant-Dubreuil -- christine.tranchant@ird.fr">
    

    <meta name="viewport" content="width=device-width,initial-scale=1">

    <link rel="stylesheet" type="text/css" media="screen" href="http://toggle.ird.fr:4000/css/bootstrap.min.css" >
    <link rel="stylesheet" type="text/css" media="screen" href="https://maxcdn.bootstrapcdn.com/font-awesome/4.4.0/css/font-awesome.min.css">
    <link rel="stylesheet" type="text/css" media="screen" href="http://toggle.ird.fr:4000/css/monokai.css">
    <link rel="stylesheet" type="text/css" media="screen" href="http://toggle.ird.fr:4000/css/stylesheet.css">
    <link rel="icon" href="http://toggle.ird.fr:4000/images/southgreenlong.png">
</head>
 <body>

      <!-- Static navbar -->
      <nav class="navbar navbar-default navbar-inverse navbar-static-top">
        <div class="container">
          <div class="navbar-header">
            <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false" aria-controls="navbar">
              <span class="sr-only"></span>
              <span class="icon-bar"></span>
              <span class="icon-bar"></span>
              <span class="icon-bar"></span>
            </button>
            <a class="navbar-brand visible-xs" href="#"></a>
          </div>
          <div id="navbar" class="navbar-collapse collapse">
			<ul id="menu">
				<li><a href="http://toggle.ird.fr:4000/">Home</a></li>
				<li><a href="#">Linux</a><span class="caret" ></span>
					<ul id="child">
						<li><a href="http://toggle.ird.fr:4000/linux/">For Dummies</a></li>
						<li><a href="http://toggle.ird.fr:4000/linuxJedi/">For Jedi</a></li>
					</ul>
				</li>
				<li><a href="http://toggle.ird.fr:4000/python/">Python</a></li>
				<li><a href="http://toggle.ird.fr:4000/perl/">Perl</a></li>
    <!--<li><a href="http://toggle.ird.fr:4000/R/">R</a></li>-->
        <li><a href="#">HPC</a><span class="caret" ></span>
					<ul>
				<li><a href="http://toggle.ird.fr:4000/HPC/">HPC</a></li>
				<li><a href="http://toggle.ird.fr:4000/Advanced_HPC/">Advanced HPC</a></li>
        </ul>
				</li><li><a href="#">Workflow</a><span class="caret" ></span>
					<ul>
						<li><a href="http://toggle.ird.fr:4000/toggle/">TOGGLe</a></li>
						<li><a href="http://toggle.ird.fr:4000/galaxy/">Galaxy</a></li>
						<li><a href="http://toggle.ird.fr:4000/gigwa/">GIGWA</a></li>
						<li><a href="http://toggle.ird.fr:4000/galaxyToggle/">Workflow Manager</a></li>
					</ul>
				</li>
                               <li><a href="http://toggle.ird.fr:4000/ngs/">NGS</a><span class="caret" ></span>
                                        <ul>
                                              <!--   <li><a href="http://toggle.ird.fr:4000/todo/">GBS</a></li>-->
                                                <li><a href="http://toggle.ird.fr:4000/rnaseq/">RNASeq</a></li>
                                                <li><a href="http://toggle.ird.fr:4000/metabarcoding/">Meta- barcoding</a></li>
						<li><a href="http://toggle.ird.fr:4000/gbsGuadeloupe/">GBS</a></li>
                                        </ul>
                                </li>
	<!--			<li><a href="#">Schools</a><span class="caret" ></span>
                                        <ul>
                                                <li><a href="http://toggle.ird.fr:4000/todo/">Musa genome</a></li>
                                                <li><a href="http://toggle.ird.fr:4000/todo/">France</a></li>
                                                <li><a href="http://toggle.ird.fr:4000/todo/">International</a></li>
                                        </ul>
                                </li>-->
				<li><a href="http://toggle.ird.fr:4000/contact/">About Us</a></li>
				<li><a id="forkme_banner" href="https://github.com/SouthGreenPlatform/trainings/">GitHub</a></li>
			</ul>
          </div>
        </div>
      </nav>
      <!--/.nav-collapse -->
    <!-- Header -->
    <div id="header_wrap" class="outer">
        <header class="inner">
          <div class="header"><img class="img-responsive" width="25%" src="http://toggle.ird.fr:4000/images/southgreenlong.png" alt="South Green Logo" /></div>
          <h2 id="project_tagline">South Green Trainings pages</h2>
        </header>
    </div>

<body>
<div id="main_content_wrap" class="outer">
	 <section id="main_content" class="inner">
<table>
  <thead>
    <tr>
      <th style="text-align: left">Description</th>
      <th style="text-align: left">Hands On Lab Exercises for Metabarcoding</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td style="text-align: left">Authors</td>
      <td style="text-align: left">Julie Orjuela (julie.orjuela@ird.fr), Alexis Dereeper (alexis.dereeper@ird.fr), Florentin Constancias (florentin.constancias@cirad.fr)</td>
    </tr>
    <tr>
      <td style="text-align: left">Creation Date</td>
      <td style="text-align: left">18/04/2018</td>
    </tr>
    <tr>
      <td style="text-align: left">Last Modified Date</td>
      <td style="text-align: left">18/04/2018</td>
    </tr>
  </tbody>
</table>

<hr />

<h3 id="summary">Summary</h3>

<!-- TOC depthFrom:2 depthTo:2 withLinks:1 updateOnSave:1 orderedList:0 -->
<ul>
  <li><a href="#practice-1">Practice 1: Obtaining an OTU table with FROGS in Galaxy</a></li>
  <li><a href="#practice-2">Practice 2: Visualizing and plotting all sample results with Phinch</a></li>
  <li><a href="#practice-3">Practice 3: Handling and visualizing OTU table using PhyloSeq R package</a></li>
  <li><a href="#links">Links</a></li>
  <li><a href="#license">License</a></li>
</ul>

<p><a name="practice-1"></a></p>
<table class="table-contact">
<tr>
<td width="25%"><img width="60%" src="http://toggle.ird.fr:4000/images/trainings-galaxy.png" alt="" />
</td>
<td width="25%"><img width="30%" src="http://toggle.ird.fr:4000/images/FROGS_logo.png" alt="" />
</td>
<td width="25%"><img width="70%" src="http://toggle.ird.fr:4000/images/phinch.png" alt="" />
</td> 
<td width="25%"><img width="70%" src="http://toggle.ird.fr:4000/images/bioconductor.png" alt="" />
</td> 
</tr>
</table>

<h3 id="practice-1--obtaining-an-otu-table-with-frogs-in-galaxy">Practice 1 : Obtaining an OTU table with FROGS in Galaxy</h3>

<p>In this training we will performed metabarcoding analysis with the FROGS pipeline in the Galaxy environment.</p>
<ul>
  <li>Connect to <a href="http://galaxy.southgreen.fr/galaxy/">Galaxy South Green</a> with formationN account.</li>
  <li>Create a new history and import 8 Metabarcoding sample datasets (paired-end fastq files) from Data library
<code class="highlighter-rouge">Galaxy_trainings_2015 =&gt; Metagenomics</code>
    <ul>
      <li>Fastq file used here are a subset of reads obtained in a metagenomic study of rice root.</li>
    </ul>
  </li>
  <li>Merge paired reads and dereplicate using the Preprocessing tool - <code class="highlighter-rouge">FROGS Pre-process</code>
    <ul>
      <li>Read size is 300 pb, expected, minimum and maximun amplicon size are 480,420,520 pb respectively. Use custom sequencing protocol.</li>
      <li>How many sequences have been overlapped?</li>
      <li>How many sequences remain after dereplication?</li>
      <li>What amplicon size is obtained in the majority of merged sequences?</li>
    </ul>
  </li>
  <li>Build Clustering using swarm - <code class="highlighter-rouge">FROGS Clustering swarm</code>
    <ul>
      <li>Use an aggregation distance of 1.</li>
      <li>The biom file shows the abundance of each cluster.</li>
      <li>The fasta file contains the cluster (OTU) representative sequences.</li>
      <li>The tsv file shows what sequences are contained in each cluster.</li>
    </ul>
  </li>
  <li>Obtain statistics about abundance of sequences in clusters - <code class="highlighter-rouge">FROGS Clusters stat</code>
    <ul>
      <li>How many clusters were obtained by swarm?</li>
      <li>How many sequences are contained in the biggest cluster?</li>
      <li>How many clusters contain only one sequence?</li>
      <li>Observe the cumulative sequence proportion by cluster size</li>
      <li>Observe cluster sharing between samples through hierarchical clustering tree</li>
    </ul>
  </li>
  <li>Remove chimera - <code class="highlighter-rouge">FROGS Remove chimera</code>
    <ul>
      <li>What proportion of clusters were kept in this step?</li>
    </ul>
  </li>
  <li>Filters OTUs on several criteria. - <code class="highlighter-rouge">FROGS Filters</code>
    <ul>
      <li>Eliminate OTUs with a low number of sequences (abundance &lt; 10) and keep OTUs present in at least two samples.</li>
      <li>How many OTUs were removed in this step?</li>
      <li>How many OTUs were removed because of low abundance?</li>
    </ul>
  </li>
  <li>Rerun statistics of clusters after filtering - <code class="highlighter-rouge">FROGS Clusters stat</code>
    <ul>
      <li>Look the effect of the cumulative proportion by cluster size.</li>
    </ul>
  </li>
  <li>Perform taxonomic affiliation of each OTU by BLAST - <code class="highlighter-rouge">FROGS Affiliation OTU</code>
    <ul>
      <li>Use the SILVA 16S database for taxonomic assignation by BLAST.</li>
      <li>How many OTU were taxonomically assigned to species?</li>
      <li>Visualize the biom file enriched with taxomonic information.</li>
    </ul>
  </li>
  <li>Obtain statistics of affiliation - <code class="highlighter-rouge">FROGS Affiliation stat</code>
    <ul>
      <li>Observe global distribution of taxonomies by sample.</li>
      <li>Look the rarefaction curve, which is a measure of samples vs diversity.</li>
    </ul>
  </li>
  <li>Retrieve a (human readable) OTU table in tsv format - <code class="highlighter-rouge">FROGS BIOM to TSV</code>
    <ul>
      <li>Download the tsv file for potential subsequent filtering.</li>
    </ul>
  </li>
</ul>

<hr />

<p><a name="practice-2"></a></p>
<h3 id="practice-2--visualizing-and-plotting-sample-results-with-phinch">Practice 2 : Visualizing and plotting sample results with Phinch</h3>
<td>Practice2 will use a specialized website to have an overview of a complete biom file. This dataset includes 24 samples of rice microbiome : 3 fields of sampling, 4 replicates, for infected and non-infected. These details were inclued in the metadata information on biom file. </td>
<ul>
  <li>Before start, retrieve the biom file containing taxomonic information of the whole rice samples.
From Galaxy, download  the file riz2.biom.txt from Data library <code class="highlighter-rouge">Galaxy_trainings_2015 =&gt; Metagenomics</code>.</li>
  <li>Connect to <a href="http://phinch.org/">Phinch</a> and import this biom file (obtained by FROGS).</li>
  <li>Select the 24 samples in ordet to analyse the whole data.</li>
  <li>Explore the <code class="highlighter-rouge">Graph Gallery</code>.</li>
  <li>Observe the diversity contained in each sample by using <code class="highlighter-rouge">Taxonomy Bar Chart</code>. Change the lineage level to family, genus or species. Change also to percentage view.</li>
  <li>Observe the <code class="highlighter-rouge">Donut Partition</code> and compare the infected with the non-infected. It combines the distribution from the 3 fields infected versus the 3 fields not infected. Can you observe some differencies? Are there some over-represented phylum or order in infected fields?</li>
</ul>

<hr />

<p><a name="practice-3"></a></p>
<h3 id="practice-3--handling-and-visualizing-otu-table-using-phyloseq-r-package">Practice 3 : Handling and visualizing OTU table using PhyloSeq R package</h3>
<td>Practice3 will be performed in the R environment using Rstudio.</td>
<ul>
  <li>From Galaxy, download  the file riz_metadata.txt from Data library <code class="highlighter-rouge">Galaxy_trainings_2015 =&gt; Metagenomics</code></li>
  <li>Download <a href="https://southgreenplatform.github.io/trainings/files/phyloseq.r">phyloseq.r</a> file and import it in Rstudio.</li>
  <li>Import ‘OTU’ table and metadata</li>
  <li>Generate rarefaction curves, filter and normalize ‘OTU’ table</li>
  <li>Visualize and alpha-diversity, beta-diversity - generate basic statistics</li>
  <li>Generate taxonomy summary at different levels</li>
</ul>

<hr />

<h3 id="links">Links</h3>
<p><a name="links"></a></p>
<ul>
  <li>Related courses : <a href="http://sepsis-omics.github.io/tutorials/modules/frogs/">Metabarcoding</a></li>
</ul>

<hr />

<h3 id="license">License</h3>
<p><a name="license"></a></p>

<div>
The resource material is licensed under the Creative Commons Attribution 4.0 International License (<a href="http://creativecommons.org/licenses/by-nc-sa/4.0/">here</a>).
<center><img width="25%" class="img-responsive" src="http://creativecommons.org.nz/wp-content/uploads/2012/05/by-nc-sa1.png" />
</center>
</div>

	</div> <!-- #main -->
</div> <!-- #main-container -->
    <!-- Placed at the end of the document so the pages load faster -->
    <script src="http://toggle.ird.fr:4000/js/jquery-1.11.3.min.js"></script>
    <script src="http://toggle.ird.fr:4000/js/bootstrap.min.js"></script>
    <script src="http://toggle.ird.fr:4000/js/homepage.js"></script>
    <script src="http://toggle.ird.fr:4000/js/libs/modernizr-2.0.6.min.js"></script>

</script>

    <!-- FOOTER  -->
    <div id="footer_wrap" class="outer">
      <footer class="inner">
        <div id="footer">
			<div class="logo-div" class="clearfix">
				<a class="logo" href="http://www.ird.fr">
					<img class="img-logo" src="http://toggle.ird.fr:4000/images/logo_ird.png" alt="IRD Logo"  width="" />
				</a>

				<a class="logo" href="http://www.cirad.fr">
					<img class="img-logo" src="http://toggle.ird.fr:4000/images/logo-cirad.png" alt="CIRAD Logo"   width="" />
				</a>

				<a class="logo" href="http://www.inra.fr">
					<img class="img-logo" src="http://toggle.ird.fr:4000/images/logo_inra.png" alt="INRA Logo"   width=""/>
				</a>

				<a class="logo" href="https://www.bioversityinternational.org/">
					<img class="img-logo" src="http://toggle.ird.fr:4000/images/logo_bioversity.png" alt="Bioversity Logo"   width=""/>
				</a>

				<a class="logo" href="https://www.supagro.fr/">
					<img class="img-logo" src="http://toggle.ird.fr:4000/images/logo_supagro.png" alt="SupAgro Logo"   width=""/>
				</a>
				
				<a class="logo" href="http://www.southgreen.fr">
					<img class="img-logo" src="http://toggle.ird.fr:4000/images/southgreenlong.png" alt="South Green Logo"   width=""/>
				</a>

			</div>
		</div>

		<div class="footer_txt">
			<p>The contents of this website are &copy;&nbsp;2019 under the terms of the Licencied under
				<a href="http://www.cecill.info/licences/Licence_CeCILL-C_V1-en.html">CeCill-C</a>
				 and <a href="/blob/master/LICENSE">GPLv3</a>.
			</p>
            <p class="webmaster">WebMaster: <a class="webmaster" target="_blank" href="https://d2gg9evh47fn9z.cloudfront.net/800px_COLOURBOX6415001.jpg">Christine</a>
            &
            <a class="webmaster" target="_blank" href="http://moziru.com/images/fluffy-clipart-silly-monster-13.png">Séb</a></p>
		</div>
      </footer>
    </div>

</body>
</html>
